### DISCUSSION TITLE
Linux support not ok for swiftformat

### DISCUSSION THREAD

**Discussion by doozMen** [airbnb/swift#197] [General]
The build binary of swiftformat at url https://github.com/calda/SwiftFormat/releases/ has support only for apple platforms

```json
{
  "schemaVersion": "1.0",
  "artifacts": {
    "swiftformat": {
      "version": "0.50-beta-4",
      "source": "https://github.com/calda/SwiftFormat/releases/tag/0.50-beta-4",
      "type": "executable",
      "variants": [
        {
          "path": "swiftformat",
          "supportedTriples": ["x86_64-apple-macosx", "arm64-apple-macosx"]
        }
      ]
    }
  }
}

```

Hence running the format tool on linux does not work but the Readme states it is compatible with linux, but unfortunately it is not.

**Comment by calda** [comment]
> the Readme states it is compatible with linux

This may be a bug in the Swift Package Index compatibility detection. I think that badge previously only said that the package supported macOS, but now (incorrectly) says it supports Linux. Would you be interested in filing a [bug report](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/issues/new/choose) on the Swift Package Index repo about this?

Regarding actually supporting Linux for this tool -- I can look in to this next time I update the SwiftFormat binary that we use.

**Reply by doozMen** [reply]
It would be so awsome if you could update the swiftformat binary to support linux

**Reply by doozMen** [reply]
As suggested I reported the bug. https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/issues/2116

But it seams not to difficult to support linux, please let me know if you do on the next release. Thanks

**Reply by finestructure** [reply]
> This may be a bug in the Swift Package Index compatibility detection

Yes, this is a problem with our compatibility detection for plugins. We're building the plugin and that just works™ on Linux 😅

As I mentioned in the issue, I'm not sure how to best fix this on our end. It'd certainly take a while to sort out. In the meantime, a quick (and rather ugly) fix would be to break the Linux build intentionally by inserting

```swift
#if os(Linux)
#error("Linux is currently not supported")
#endif
```

somewhere in the source (or even the `Package.swift` manifest). Would that work for you?

**Reply by calda** [reply]
good idea, thanks for the suggestion!

**Reply by calda** [reply]
@finestructure, I made this change in https://github.com/airbnb/swift/pull/206 but it seems to have broken our package's pack on Swift Package Index: https://swiftpackageindex.com/airbnb/swift

**Reply by calda** [reply]
Here's a bug report for that issue: https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/issues/2117

**Reply by finestructure** [reply]
As fas as I can see everything is working as expected - see my comment over in the issue! Not sure what went wrong there but it seems to be ok now :)

**Reply by rswave** [reply]
This is causes github's dependabot to fail. Any workaround for that?

```
updater | Computing version for https://github.com/airbnb/lottie-spm
updater | error: Invalid manifest (compiled with: ["/opt/swift/usr/bin/swiftc", "-vfsoverlay", "/tmp/TemporaryDirectory.WBnePc/vfs.yaml", "-L", "/opt/swift/usr/lib/swift/pm/ManifestAPI", "-lPackageDescription", "-Xlinker", "-rpath", "-Xlinker", "/opt/swift/usr/lib/swift/pm/ManifestAPI", "-swift-version", "5", "-I", "/opt/swift/usr/lib/swift/pm/ManifestAPI", "-package-description-version", "5.6.0", "/Package.swift", "-Xfrontend", "-disable-implicit-concurrency-module-import", "-Xfrontend", "-disable-implicit-string-processing-module-import", "-o", "/tmp/TemporaryDirectory.dMTZCA/lottie-spm-manifest"])
updater | error: fatalError
updater | /Package.swift:46:8: error: Linux is currently not supported
updater | #error("Linux is currently not supported")
updater |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ in https://github.com/airbnb/lottie-spm
updater | 2024/05/20 22:54:30 ERROR <job_830742886> /home/dependabot/common/lib/dependabot/shared_helpers.rb:432:in `run_shell_command'
updater | 2024/05/20 22:54:30 ERROR <job_830742886> /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/sorbet-runtime-0.5.11353/lib/types/private/methods/call_validation.rb:167:in `bind_call'
```

**Reply by AllDmeat** [reply]
Same issue with Dependabot. That's a security issue in my company's project, because we can't receive automatic security updates no more.

**Reply by calda** [reply]
Thanks -- we'll remove the `#error` from our Package.swift: https://github.com/airbnb/swift/pull/278

**Comment by doozMen** [comment]
I fixed this issue for now for us by depending not on the binary of swiftformat but on the release. https://github.com/doozMen/swift/releases/tag/1.0.2 points to https://github.com/calda/SwiftFormat/releases/tag/0.51-beta-3. This is a workaround as now our ci pipeline is slower but at least it can run on linux ...

**Reply by calda** [reply]
Good idea, we could use `#if os(Linux)` to build SwiftFormat from source instead of using the binary dependency. 

How did you handle SwiftLint? That [binary dependency](https://github.com/realm/SwiftLint/releases/download/0.48.0/SwiftLintBinary-macos.artifactbundle.zip) also appears to only support macOS.

**Reply by doozMen** [reply]
I made manual builds now for linux and updated the swiftformat and swiftlint bundle artifact. Bot on my machine with `aarch64-linux-gnu`, `arm64-apple-darwin22.1.0` and our pipeline running with docker `aarch64-linux-gnu` and ubuntu 18 this now works.

https://github.com/doozMen/swift/releases/tag/1.0.6

The problem is that I did not automate the building process of the artefact bundle. I would need some help with that as you would need to make builds on 3 machines with the above architecture. In essence the problem with both swiftlint and swiftformat binaries is that they are build for linux with fat binaries that do not work on arm64 as explained https://www.polpiella.dev/binary-targets-in-modern-swift-packages#fat-binaries-and-frameworks. I hope both swiftformat and swiftlint would include in there future releases an artifactbundle like the above and include the checksum needed.

**Reply by doozMen** [reply]
> Good idea, we could use `#if os(Linux)` to build SwiftFormat from source instead of using the binary dependency. 
> 
> 
> 
> How did you handle SwiftLint? That [binary dependency](https://github.com/realm/SwiftLint/releases/download/0.48.0/SwiftLintBinary-macos.artifactbundle.zip) also appears to only support macOS.

I Also build swiftlint from source. But it is also possible to add multiple target triples to the artefactbundle info.plist like I did in my comment. That way it works on Linux with a binary target. Just requires building on 3 machines. 
That last step I have not automated unfortunately
